lldb/source/Plugins/OperatingSystem/Swift/OperatingSystemSwift.cpp

```diff
@@ -0,0 +1,164 @@
+
+#if LLDB_ENABLE_SWIFT
+
+#include "OperatingSystemSwift.h"
+
+#include "Plugins/Process/Utility/ThreadMemory.h"
+#include "lldb/Core/Debugger.h"
+#include "lldb/Core/Module.h"
+#include "lldb/Core/PluginManager.h"
+#include "lldb/Target/Process.h"
+#include "lldb/Target/Thread.h"
+#include "lldb/Target/ThreadList.h"
+#include "lldb/Utility/LLDBLog.h"
+#include "lldb/Utility/StructuredData.h"
+
+#include "swift/Threading/ThreadLocalStorage.h"
+
+#include <memory>
+
+using namespace lldb;
+using namespace lldb_private;
+
+LLDB_PLUGIN_DEFINE(OperatingSystemSwift)
+
+void OperatingSystemSwift::Initialize() {
+  PluginManager::RegisterPlugin(GetPluginNameStatic(),
+                                GetPluginDescriptionStatic(), CreateInstance,
+                                nullptr);
+}
+
+void OperatingSystemSwift::Terminate() {
+  PluginManager::UnregisterPlugin(CreateInstance);
+}
+
+OperatingSystem *OperatingSystemSwift::CreateInstance(Process *process,
+                                                      bool force) {
+  if (!process)
+    return nullptr;
+
+  /// Heuristic to only load the plugin for swift programs.
+  FileSpec swift_module_name = FileSpec("libswiftCore.dylib");
+  if (process->GetTarget().GetImages().FindFirstModule(swift_module_name))
+    return new OperatingSystemSwift(*process);
+  return nullptr;
+}
+
+llvm::StringRef OperatingSystemSwift::GetPluginDescriptionStatic() {
+  return "Operating system plug-in converting Swift Tasks into Threads.";
+}
+
+OperatingSystemSwift::~OperatingSystemSwift() = default;
+
+OperatingSystemSwift::OperatingSystemSwift(lldb_private::Process &process)
+    : OperatingSystem(&process) {
+  size_t ptr_size = process.GetAddressByteSize();
+  // Offset of the Task pointer in a Thread's local storage.
+  m_task_ptr_offset_in_tls =
+      swift::tls_get_key(swift::tls_key::concurrency_task) * ptr_size;
+  // Offset of a Task ID inside a Task data structure, guaranteed by the ABI.
+  // See Job in swift/RemoteInspection/RuntimeInternals.h.
+  m_task_id_offset = 4 * ptr_size + 4;
+}
+
+bool OperatingSystemSwift::UpdateThreadList(ThreadList &old_thread_list,
+                                            ThreadList &core_thread_list,
+                                            ThreadList &new_thread_list) {
+  Log *log = GetLog(LLDBLog::OS);
+  LLDB_LOG(log, "OperatingSystemSwift: Updating thread list");
+
+  for (const ThreadSP &real_thread : core_thread_list.Threads()) {
+    std::optional<uint64_t> task_id = FindTaskId(*real_thread);
+
+    // If this is not a thread running a Task, add it to the list as is.
+    if (!task_id) {
+      new_thread_list.AddThread(real_thread);
+      LLDB_LOGF(log,
+                "OperatingSystemSwift: thread %" PRIx64
+                " is not executing a Task",
+                real_thread->GetID());
+      continue;
+    }
+
+    // Mask higher bits to avoid conflicts with core thread IDs.
+    uint64_t masked_task_id = 0xdeadbeef00000000 | *task_id;
```

Comment:
This will be the visible "TID" for running tasks, right?  And what I'd have to type if I wanted to set a "task specific" breakpoint?  If so, we might not want it to be deadbeef.  That's cute if you know the reference, but otherwise somewhat off-putting...

Reply:
Oh yeah, I'm actually looking for suggestions here, if we should mask it at all and what to use if we do
Deadbeef was just something to help me grep the logs as I debugged.